generate/main.go

```diff
@@ -40,44 +37,156 @@ func main() {
 		ipv6File = local
 	}
 
-	ipv6CidrToAsnMap := readMappingFile(ipv6File)
-	f, err := os.Create(ipv6OutputFile)
+	networks := readMappingFile(ipv6File)
+
+	// Keep fixing and optimizing until we get a stable configuration.
+	for {
+		before := slices.Clone(networks)
+		// Ensure the networks are sorted, first smallest start then biggest end.
+		slices.SortStableFunc(networks, func(a, b entry) int {
+			switch {
+			case a.start < b.start:
+				return -1
+			case a.start == b.start:
+				switch {
+				case a.end < b.end:
+					return 1
+				case a.end == b.end:
+					return 0
+				default:
+					return -1
+				}
+			default:
+				return 1
+			}
+		})
+
+		// Merge adjacent ranges
+		{
+			var new []entry
+			var old entry
+			for _, n := range networks {
+				if n.start <= old.end || n.start <= old.end+1 {
+					// mergeable
+					if n.asn == old.asn {
+						// merge
+						if n.end > old.end {
+							old.end = n.end
+						}
+						continue
+					}
+					// We have an overlap of different networks :'(
+					// Real example:
+					//	2403:8080::	2403:8080:ffff:ffff:ffff:ffff:ffff:ffff	17964	CN	DXTNET Beijing Dian-Xin-Tong Network Technologies Co., Ltd.
+					//	2403:8080:101::	2403:8080:101:ffff:ffff:ffff:ffff:ffff	4847	CN	CNIX-AP China Networks Inter-Exchange
+					// Split the networks.
+					if n.start < old.start {
+						n, old = old, n
+					}
+					switch {
+					case n.start > old.start && n.end < old.end:
+						// |   old   |
+						//   | new |
+						new = append(new, entry{old.start, n.start - 1, old.asn}, n)
+						old.start = n.end + 1
+					case n.start > old.start && n.end == old.end:
+						// |   old   |
+						//     | new |
+						fallthrough
+					case n.start > old.start && n.end > old.end:
+						// |   old   |
+						//       | new |
+						new = append(new, entry{old.start, n.start - 1, old.asn})
+						old = n
+					case n.start == old.start && n.end > old.end:
+						// |   old   |
+						// |    new    |
+						n, old = old, n
+						fallthrough
+					case n.start == old.start && n.end < old.end:
+						// |   old   |
+						// | new |
+						new = append(new, n)
+						old.start = n.end + 1
```

Comment:
This heuristic may lead to the following case of equal networks.

in general you want to accept the most specific network you have, so if you have a general /23 and two more specific /24s's taking up it's two halves you want to take the two more specific definitions and ignore the fully masked broader one.

Here, by cutting down the broader network's assumed space when you see a more specific one, you may end up in the next case where you drop the subsequent specific definition in favor of the truncated general definition.

Reply:
This dataset is not CIDR aligned, this were a bug in the previous implementation. We can't easily know which one is the "biggest one" here.

I found a bug but I don't know if it is the same.
For example assume you have:
- `0b10/1` as `1`
- `0b10/2` as `2`
- `0b11/2` as `3`
This represent:
```go
|   1   |
| 2 | 3 |
```
First the sort step will sort it that way:
- `0b1000000000000000000000000000000000000000000000000000000000000000`, `0b111111111111111111111111111111111111111111111111111111111111111` as `1`
- `0b1000000000000000000000000000000000000000000000000000000000000000`, `0b1011111111111111111111111111111111111111111111111111111111111111` as `2`
- `0b1100000000000000000000000000000000000000000000000000000000000000`, `0b1111111111111111111111111111111111111111111111111111111111111111` as `3`

So first it loads `1` into `old`.
Then it compare `2` with `1` and enters:
```go
						// |   old   |
						// | new |
```
At this point `2` is pushed onto the resulting slice and `1` is truncated to become upper the half.
Then it compare truncated `1` with `3` and reach:
```go
						// |   old   |
						// |   new   |
						// ¯\_(ツ)_/¯ here we are kinda fucked
						// in theory we could try merging boths with the next elements and keep the smallest one
						// but that is tricky since the next elements could also need the same treatment.
						// just drop new and too bad.
```
And drops `3`.
So we get:
```
| 2 | 1 |
```
---
I'll change:
```go
						// |   old   |
						// |   new   |
```
to swap old with new, since the list is sorted we should more often than not be correct (it wont be correct if the biggest element aligned would change after being merged, I don't think that would even happens in today's dataset).
So after this fix we will get:
```
| 2 | 3 |
```

Is this what you had in mind ?